File: Algorithms/Sorting_Algorithms/Quick_Sort/C/Quick_Sort.c
//Program in C to demonstrate quick sort algorithm

#include<stdio.h>

//Utility function to swap the numbers using call by reference
int swap(int *x,int *y)
{
	int temp=*x;
	*x=*y;
	*y=temp;
}

//Iterative utility function which does partion among the array elements
//Partition is done with respect to a pivot element
int partition(int array[],int lb,int ub)
{
	int start=lb;
	int end=ub;
	int pivot=array[lb];
	while(start<end)
	{
		//to the left of pivot,elements are less than or equal to it
		while(array[start]<=pivot)
		{
			start++;
		}
		//to the right of pivot,elements are greater than it
		while(array[end]>pivot)
		{
			end--;
		}
		//swap start with end only if start is less than end
		if(start<end)
		{
			swap(&array[start],&array[end]);
		}
	}
	//otherwise swap pivot with end
	swap(&array[lb],&array[end]);
	return end;
}

//Recursive utility function which returns the array sorted in ascending order
int quickSort(int array[],int lb,int ub)
{
	if(lb<ub)
	{
		//loc stores location of end so that we can break the array from pivot and proceed further
		int loc=partition(array,lb,ub);
		//recursive calls that sort the array further
		quickSort(array,lb,loc-1);
		quickSort(array,loc+1,ub);
	}
}

//Utility function to print the array
int printArray(int array[],int lb,int ub)
{
	printf(" The required sorted array :\n");
	for(int i=lb;i<=ub;i++)
	{
		printf(" %d ",array[i]);
	}
}

//Driver program to check the above functions
int main()
{
	int array[]={33,0,5,-9,227,84,2,-17};
	int lb=0,ub=7;
	quickSort(array,lb,ub);
	printArray(array,lb,ub);
	return 0;
}

File: Algorithms/Sorting_Algorithms/Quick_Sort/Java/Quick_Sort.java
package sorting;

public class QuickSortMain {

	public static void main(String[] args) {
		QuickSort obj = new QuickSort();
		int a[] = {7,3,1,-9,-4,6,11,24,55,2,0};
		
		System.out.println("");
		obj.quicksort(a,0,a.length-1);
		obj.printArray(a);

	}

}
class QuickSort{
	void quicksort(int arr[],int start, int end) {
		if(start<end) {
			int pivot=partition(arr,start,end);
			quicksort(arr,start,pivot-1);
			quicksort(arr,pivot+1,end);
		}
	}
	
	int partition(int arr[],int p,int q) {
		int i=p-1;
		for(int j=p;j<=q;j++) {
			if(arr[j]<=arr[q]) {
				i++;
				int temp=arr[j];
				arr[j]=arr[i];
				arr[i]=temp;
			}
		}
		return i;
	}
	//printing sorted array
	void printArray(int arr[]) {
		for(int i=0;i<arr.length;i++) {
			System.out.print(arr[i]+" ");
		}
	}
}
